### PR TITLE
chore: backport apify-js commits

### PR DESCRIPTION
**Pull request by barjin**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -17,6 +17,40 @@
...
+- reclaimed request
+
+If we don't detect one of those actions in last 5 minutes, and we have some
+requests in the `inProgress` cache, we try to reset the state. We can override
+this limit via `APIFY_INTERNAL_TIMEOUT` env var.
+
+This should finally resolve the 0 concurrency bug, as it was always about
+stuck requests in the `inProgress` cache.
+
 2.2.2 / 2022/02/xx
```
Comment (B4nan) [inline CHANGELOG.md:54]: let's fix this too

File: docs/guides/avoid_blocking.mdx [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+Browser fingerprint is a collection of browser attributes and significant features that can show if your browser is a bot or a real user. Moreover, even most browsers have these unique features that allow the website to track the browser even within different IP addresses. This is the main reason why scrapers should change browser fingerprints while doing browser-based scraping. In addition, it should reduce blocking significantly.
+
+Changing browser fingerprints can be a tedious job. Luckily, Apify SDK provides this feature out of the box with zero configuration necessary. Let's take a look at how it is done.
+
+ Changing browser fingerprints is available in `PuppeteerCrawler` and `PlaywrightCrawler`. You have to pass the `useFingerprints` option to the `browserPoolOptions`.
+
+ ```javascript
+const crawler = new Apify.PlaywrightCrawler({
```
Comment (B4nan) [inline docs/guides/avoid_blocking.mdx:14]:
```suggestion
const crawler = new PlaywrightCrawler({
```

File: docs/guides/avoid_blocking.mdx [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    browserPoolOptions: {
+        useFingerprints: true,
+    },
+})
+ ```
+Now, it is all set. The fingerprints are going to be generated for the default browser and the operating system. The Crawler can have the generation alghoritm customized to reflect particular browser version and many more. Let's take a look at the example bellow:
+
+ ```javascript
+const crawler = new Apify.PlaywrightCrawler({
```
Comment (B4nan) [inline docs/guides/avoid_blocking.mdx:23]:
```suggestion
const crawler = new PlaywrightCrawler({
```

File: docs/guides/avoid_blocking.mdx [hunk 4]
```diff
@@ -0,0 +1,44 @@
+---
+id: avoid-blocking
+title: Avoid getting blocked
+---
+
+A Scraper might get blocked for numerous reasons. Let's narrow it down to two main reasons. The first one is a bad or blocked IP address. This topic is covered in the [proxy management guide](proxy_management.mdx). The second reason we will explore more is [browser fingerprints](https://pixelprivacy.com/resources/browser-fingerprinting/) or signatures.
+Browser fingerprint is a collection of browser attributes and significant features that can show if your browser is a bot or a real user. Moreover, even most browsers have these unique features that allow the website to track the browser even within different IP addresses. This is the main reason why scrapers should change browser fingerprints while doing browser-based scraping. In addition, it should reduce blocking significantly.
+
+Changing browser fingerprints can be a tedious job. Luckily, Apify SDK provides this feature out of the box with zero configuration necessary. Let's take a look at how it is done.
```
Comment (B4nan) [inline docs/guides/avoid_blocking.mdx:9]:
```suggestion
Changing browser fingerprints can be a tedious job. Luckily, Crawlee provides this feature out of the box with zero configuration necessary. Let's take a look at how it is done.
```

this is repeated many times in current docs, would be good if you adjust the rest in a separate PR

File: packages/core/src/storages/key_value_store.ts [hunk 5]
```diff
@@ -202,8 +202,9 @@ export class KeyValueStore {
             validator: ow.isValid(k, ow.string.matches(KEY_VALUE_STORE_KEY_REGEX)),
             message: 'The "key" argument must be at most 256 characters long and only contain the following characters: a-zA-Z0-9!-_.\'()',
         })));
-        if (options.contentType && !ow.isValid(value, ow.any(ow.string, ow.buffer))) {
-            throw new ArgumentError('The "value" parameter must be a String or Buffer when "options.contentType" is specified.', this.setValue);
+        // @ts-expect-error `.pipe` duck typing
+        if (options.contentType && !(ow.isValid(value, ow.any(ow.string, ow.buffer)) || (ow.isValid(value, ow.object) && typeof value?.pipe === 'function'))) {
```
Comment (B4nan) [inline packages/core/src/storages/key_value_store.ts:206]: `@ts-ignore/expect-error` should be rather last resort, better to cast just the problematic part and keep the type checks working for the rest of the line. i am ok with those in tests, but not so much in the actual codebase

```suggestion
        if (options.contentType && !(ow.isValid(value, ow.any(ow.string, ow.buffer)) || (ow.isValid(value, ow.object) && typeof (value as Dictionary)?.pipe === 'function'))) {
```

File: packages/core/src/storages/request_queue.ts [hunk 6]
```diff
@@ -570,12 +580,32 @@ export class RequestQueue {
...
         if (this.queueHeadDict.length() > 0 || this.inProgressCount() > 0) return false;
 
         const isHeadConsistent = await this._ensureHeadIsNonEmpty(true);
         return isHeadConsistent && this.queueHeadDict.length() === 0 && this.inProgressCount() === 0;
     }
 
+    /**
+     * @internal
+     */
+    _reset() {
```
Comment (B4nan) [inline packages/core/src/storages/request_queue.ts:598]: this should be private and we can remove the annotation
```suggestion
    private _reset() {
```

File: packages/playwright-crawler/src/internals/playwright-launcher.ts [hunk 7]
```diff
@@ -47,6 +47,19 @@ export interface PlaywrightLaunchContext extends BrowserLaunchContext<LaunchOpti
      */
     useChrome?: boolean;
 
+    /**
+    * With this option selected, all pages will be opened in a new incognito browser context.
+    * This means they will not share cookies nor cache and their resources will not be throttled by one another.
+    * @default false
+    */
+     useIncognitoPages?: boolean;
+     /**
```
Comment (B4nan) [inline packages/playwright-crawler/src/internals/playwright-launcher.ts:56]:
```suggestion
     useIncognitoPages?: boolean;

     /**
```

File: packages/puppeteer-crawler/src/internals/puppeteer-launcher.ts [hunk 8]
```diff
@@ -2,11 +2,11 @@ import ow from 'ow';
 import { Browser } from 'puppeteer';
 import { PuppeteerPlugin } from '@crawlee/browser-pool';
 import { BrowserLaunchContext, BrowserLauncher } from '@crawlee/browser';
+import log from '@apify/log';
 import { applyStealthToBrowser, StealthOptions } from './stealth';
 
 /**
  * The default user agent used by `Actor.launchPuppeteer`.
- * Last updated on 2020-05-22.
```
Comment (B4nan) [inline packages/puppeteer-crawler/src/internals/puppeteer-launcher.ts:9]: why was this removed? seems to be still present in the js repo

https://github.com/apify/apify-js/blob/master/src/constants.js#L5

File: packages/puppeteer-crawler/src/internals/puppeteer-launcher.ts [hunk 9]
```diff
@@ -77,15 +77,24 @@ export interface PuppeteerLaunchContext extends BrowserLaunchContext<PuppeteerPl
      */
     launcher?: unknown;
 
+    /**
+    * With this option selected, all pages will be opened in a new incognito browser context.
+    * This means they will not share cookies nor cache and their resources will not be throttled by one another.
+    * @default false
+    */
```
Comment (B4nan) [inline packages/puppeteer-crawler/src/internals/puppeteer-launcher.ts:84]: bad indent
```suggestion
     * With this option selected, all pages will be opened in a new incognito browser context.
     * This means they will not share cookies nor cache and their resources will not be throttled by one another.
     * @default false
     */
```

File: packages/puppeteer-crawler/src/internals/puppeteer-launcher.ts [hunk 10]
```diff
@@ -77,15 +77,24 @@ export interface PuppeteerLaunchContext extends BrowserLaunchContext<PuppeteerPl
...
+    useIncognitoPages?: boolean;
+
     /**
      * This setting hides most of the known properties that identify headless Chrome and makes it nearly undetectable.
      * It is recommended to use it together with the `useChrome` set to `true`.
+     * @deprecated
```
Comment (B4nan) [inline packages/puppeteer-crawler/src/internals/puppeteer-launcher.ts:90]: just a head up, this will be removed in crawlee altogether (but lets keep it like this for now)

File: packages/utils/src/internals/extract-urls.ts [hunk 11]
```diff
@@ -42,7 +43,9 @@ export async function downloadListOfUrls(options: DownloadListOfUrlsOptions): Pr
         fixedUrl = `${match[1]}/gviz/tq?tqx=out:csv`;
     }
 
-    const { body: string } = await gotScraping({ url: fixedUrl, encoding });
+    // @ts-expect-error https://github.com/apify/got-scraping/issues/66
```
Comment (B4nan) [inline packages/utils/src/internals/extract-urls.ts:46]: here on the other hand is the `@ts-expect-error` exactly what we want 👍

File: test/apify/storages/request_queue_reset.ts [hunk 12]
```diff
@@ -0,0 +1,32 @@
+// $ npm run build && ts-node test/apify/storages/request_queue_reset.ts
+
+import { Actor } from 'apify';
+import log from '@apify/log';
+import { purgeLocalStorage } from 'crawlee';
```
Comment (B4nan) [inline test/apify/storages/request_queue_reset.ts:5]:
```suggestion
import { purgeLocalStorage } from '@crawlee/utils';
```

File: test/apify/storages/request_queue_reset.ts [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+        requestHandler: async (ctx) => {
+            log.info(ctx.request.id);
+        },
+    });
+
+    await crawler.run();
+}).then(
+    () => {},
+    () => {},
+);
```
Comment (B4nan) [inline test/apify/storages/request_queue_reset.ts:32]: that's because of the eslint rule? you can just prefix this with `void` to let it know its ok, or a simple `.then()` does the same job

but as said above, this should be an e2e test, so it should be ESM module, not TS file, and we can use top level await in there

File: test/apify/storages/request_queue_reset.ts [hunk 14]
```diff
@@ -0,0 +1,32 @@
+// $ npm run build && ts-node test/apify/storages/request_queue_reset.ts
```
Comment (B4nan) [inline test/apify/storages/request_queue_reset.ts:1]: this one should be an e2e test, so please move it to `test/e2e/test-request-queue`

(we might want to revisit the naming, i dont like the `test-` prefix, but lets be consistent for now)

File: test/apify/storages/request_queue_reset.ts [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+
+import { Actor } from 'apify';
+import log from '@apify/log';
+import { purgeLocalStorage } from 'crawlee';
+import { CheerioCrawler } from '@crawlee/cheerio';
+
+// RequestQueue auto-reset when stuck with requests in progress
+Actor.main(async () => {
+    process.env.APIFY_INTERNAL_TIMEOUT = '30000'; // 30s
```
Comment (B4nan) [inline test/apify/storages/request_queue_reset.ts:10]: the `APIFY_INTERNAL_TIMEOUT` env var should be renamed to `CRAWLEE_INTERNAL_TIMEOUT`

**Review by B4nan — CHANGES_REQUESTED** [review]

**Comment by B4nan** [comment]
And please backport this one too (browser pool is also part of the monorepo now, rest of the commits should be there already)

https://github.com/apify/browser-pool/commit/9330fe09c82b4a483313bf3c1b99e20eb1368022

File: test/e2e/test-web-default/test.mjs [hunk 16]
```diff
@@ -1,6 +1,6 @@
 import { getStats, getDatasetItems, run, expect, validateDataset } from '../tools.mjs';
 
-await run(import.meta.url, 'web-scraper', {
+await run(import.meta.url, 'cheerio-scraper', {
```
[the remark this answers is not in the document]
Comment (barjin) [reply test/e2e/test-web-default/test.mjs]: absolutely, fixed now

**Review by B4nan — APPROVED** [review]
